librad/src/git/header.rs

```diff
@@ -0,0 +1,187 @@
+                writeln!(f, "git-upload-pack {}\0host={}\0", self.repo, self.peer)
+            },
+            GitService::ReceivePackLs => writeln!(
+                f,
+                "git-receive-pack {}\0host={}\0ls\0",
+                self.repo, self.peer
+            ),
+            GitService::ReceivePack => {
+                writeln!(f, "git-receive-pack {}\0host={}\0", self.repo, self.peer)
+            },
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum ParseError {
+    #[error("Missing service")]
+    MissingService,
+
+    #[error("Invalid service: {0}. Must be one of `git-upload-pack` or `git-receive-pack`")]
+    InvalidService(String),
+
+    #[error("Missing repo")]
+    MissingRepo,
+
+    #[error("Invalid repo")]
+    InvalidRepo(#[from] uri::rad_urn::ParseError),
+
+    #[error("Missing host")]
+    MissingHost,
+
+    #[error("Malformed host. Must be a PeerId")]
+    MalformedHost(#[from] peer::conversion::Error),
+
+    #[error("Invalid mode: `{0}`. Must be `ls`, or absent")]
+    InvalidMode(String),
+}
+
+impl FromStr for Header {
+    type Err = ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut parts = s.split(|c| c == ' ' || c == '\0');
+
+        let service = parts.next().ok_or_else(|| ParseError::MissingService)?;
+        let repo = parts
+            .next()
+            .ok_or_else(|| ParseError::MissingRepo)
+            .and_then(|repo| repo.parse::<RadUrn>().map_err(|e| e.into()))?;
+        let peer = parts
+            .next()
+            .and_then(|peer| peer.strip_prefix("host="))
+            .ok_or_else(|| ParseError::MissingHost)
+            .and_then(|peer| peer.parse::<PeerId>().map_err(|e| e.into()))?;
+        let mode = parts.next().unwrap_or("");
+
+        let service = match service {
+            "git-upload-pack" => match mode {
+                "ls" => Ok(GitService::UploadPackLs),
+                "" | "\n" => Ok(GitService::UploadPack),
+                _ => Err(ParseError::InvalidMode(mode.to_owned())),
+            },
+
+            "git-receive-pack" => match mode {
+                "ls" => Ok(GitService::ReceivePackLs),
+                "" | "\n" => Ok(GitService::ReceivePack),
+                _ => Err(ParseError::InvalidMode(mode.to_owned())),
+            },
+
+            unknown => Err(ParseError::InvalidService(unknown.to_owned())),
+        }?;
+
+        Ok(Self::new(service, repo, peer))
+    }
+}
+
+pub struct Service(pub GitService);
+
+impl Debug for Service {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_tuple("Service")
+            .field(match self.0 {
+                GitService::UploadPackLs => &"UploadPackLs",
+                GitService::UploadPack => &"UploadPack",
+                GitService::ReceivePackLs => &"ReceivePackLs",
+                GitService::ReceivePack => &"ReceivePack",
+            })
+            .finish()
+    }
+}
+
+impl PartialEq for Service {
```

Comment:
They don't have `PartialEq` impl :scream: I think we need to start making PRs to `git2` :grimacing:

Reply:
Ya, but for more serious reasons... :disappointed: